uniffi_bindgen/src/bindings/swift/templates/CallbackInterfaceTemplate.swift

```diff
@@ -0,0 +1,121 @@
+    {%- else -%}
+    {%- endmatch %}
+    {% endfor %}
+}
+
+{% let canonical_type_name = cbi.type_().canonical_name()|class_name_swift %}
+{% let callback_internals = format!("{}Internals", canonical_type_name) -%}
+{% let callback_interface_impl = format!("{}FFI", canonical_type_name) -%}
+
+let {{ callback_interface_impl }} : ForeignCallback = 
+    { (handle: UInt64, method: Int32, args: RustBuffer) -> RustBuffer in
+
+        {% for meth in cbi.methods() -%}
+    {% let method_name = format!("invoke_{}", meth.name())|fn_name_swift %}
+
+    func {{ method_name }}(_ swiftCallbackInterface: {{ type_name }}, _ args: RustBuffer) throws -> RustBuffer {
+        defer { args.deallocate() }
+        {#- Unpacking args from the RustBuffer #}
+            {%- if meth.arguments().len() != 0 -%}
+            {#- Calling the concrete callback object #}
+            
+            let reader = Reader(data: Data(rustBuffer: args))
+            let result = swiftCallbackInterface.{{ meth.name()|fn_name_swift }}(
+                    {% for arg in meth.arguments() -%}
+                    {{ arg.name() }}: try {{ "reader"|read_swift(arg.type_()) }}
+                    {%- if !loop.last %}, {% endif %}
+                    {% endfor -%}
+                )
+            {% else %}
+            let result = swiftCallbackInterface.{{ meth.name()|fn_name_swift }}()
+            {% endif -%}
+            
+        {#- Packing up the return value into a RustBuffer #}
+                {%- match meth.return_type() -%}
+                {%- when Some with (return_type) -%}
+                let writer = Writer()
+                result.write(into: writer)
+                return RustBuffer(bytes: writer.bytes)
+                {%- else -%}
+                return RustBuffer()
+                {% endmatch -%}
+                // TODO catch errors and report them back to Rust. 
+                // https://github.com/mozilla/uniffi-rs/issues/351
+        
+    }
+    {% endfor %}
+
+        return {{ callback_internals }}.handleMap.callWithResult(handle: handle) { cb -> RustBuffer in 
+            switch method {
+                case IDX_CALLBACK_FREE: return {{ callback_internals }}.drop(handle: handle)
+                {% for meth in cbi.methods() -%}
+                {% let method_name = format!("invoke_{}", meth.name())|fn_name_swift -%}
+                case {{ loop.index }}: return try! {{ method_name }}(cb, args)
+                {% endfor %}
+                // This should never happen, because an out of bounds method index won't
+                // ever be used. Once we can catch errors, we should return an InternalError.
+                // https://github.com/mozilla/uniffi-rs/issues/351
+                default: return RustBuffer()
+            }
+        }
+    }
+
+// This erasure is probably not needed and can be replaced by 
+// a better implementation of _{{ callback_internals }}
+private class {{ type_name }}Erased: {{ type_name }} {
+    
```

Comment:
A type erased version of callback_internals, this is needed since functions don't accept protocols with associated types or generics. There is probably a nicer way of doing this.

Reply:
I wonder if this should be the class which implements `ViaFfi`, in the same way `Bool` does.

This would make passing around callback interface objects as collections, maps or options Just Work™.

I tried to get `{{ type_name }}` to provide an implementation of the protocol, but Swift stopped me in the two places I tried:

 * you can't add an extension to a protocol, because it looks like inheritance.
 * you can't add refine a private protocol.